Add initial servicenow plugin

EdwardDowling opened the pull request:
Part of the changes for #15366

EdwardDowling committed:
Add initial servicenow plugin

tigrato commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -5327,6 +5329,17 @@ message PluginOpsgenieAccessSettings {
+message PluginServicenowSettings {
+  option (gogoproto.equal) = true;
+
+  // Servicenow API endpoint.
+  string api_endpoint = 1 [(gogoproto.jsontag) = "api_endpoint,omitempty"];
+  // Servicenow API username.
+  string username = 2 [(gogoproto.jsontag) = "username,omitempty"];
+  // Servicenow API password.
+  string password = 3 [(gogoproto.jsontag) = "password,omitempty"];
```
nit: godoc style comments

```suggestion
  // ApiEndpoint is the target Servicenow API.
  string api_endpoint = 1 [(gogoproto.jsontag) = "api_endpoint,omitempty"];
  // Username is the Servicenow API username.
  string username = 2 [(gogoproto.jsontag) = "username,omitempty"];
  // Password is the Servicenow API password.
  string password = 3 [(gogoproto.jsontag) = "password,omitempty"];
```

tigrato commented on api/types/constants.go:
```diff
@@ -661,6 +661,10 @@ const (
 	ReqAnnotationSchedulesLabel = "/schedules"
 	// ReqAnnotationNotifyServicesLabel is the request annotation key at which notify services are stored for access plugins.
 	ReqAnnotationNotifyServicesLabel = "/notify-services"
+	// ReqAnnotationApprovedCloseCode is the request annotation key at which the servicenow close code for approved access requests is stored.
```
Can you please describe a bit what's the usecase for these labels and why they are required?
It's not clear to me what they are used for when reading the comment.

tigrato commented on integrations/access/servicenow/types.go:
```diff
@@ -38,6 +52,11 @@ type Incident struct {
 }
 
+const (
+	ResolutionStateResolved = "6"
```
nit: godoc
Also, please identify where these codes came from. Link to the API docs or any other references

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,591 @@
+		return nil
+	}
+
+	errors := make([]error, 0, len(slice))
```
It should be ok, but we are being really pessimistic here 😅

tigrato commented on integrations/access/servicenow/plugindata.go:
```diff
@@ -0,0 +1,72 @@
+	}
+	if str := dataMap["created"]; str != "" {
+		var created int64
+		fmt.Sscanf(dataMap["created"], "%d", &created)
```
should we check for errors returned by `fmt.Sscanf` or use `strconv.ParseInt`. 
I don't recommend `fmt.Sscanf` because it ignores any spaces existent in the `created` data.
Check out: https://go.dev/play/p/vKD_wgyI7TX

tigrato commented on integrations/access/servicenow/config.go:
```diff
@@ -0,0 +1,110 @@
+
+// LoadServicenowConfig reads the config file, initializes a new ServicenowConfig struct object, and returns it.
+// Optionally returns an error if the file is not readable, or if file format is invalid.
+func LoadServicenowConfig(filepath string) (*Config, error) {
```
This code doesn't seem to be used. Was it used for tests?
Do we configure our plugins using toml or is it used by `teleport-plugins` repo

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,591 @@
+		return nil
+	}
+
+	onCallUsers := []string{}
+	for _, scheduleName := range serviceNames {
+		respondersResult, err := a.servicenow.GetOnCall(ctx, scheduleName)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		onCallUsers = append(onCallUsers, respondersResult...)
+	}
+
+	userIsOnCall := false
+	for _, user := range onCallUsers {
+		if req.GetUser() == user {
+			userIsOnCall = true
+		}
+	}
```
Could we extract this to a function?
Also, if we invert the check for `if !userIsOnCall`, then we could remove one indentation level

EdwardDowling committed:
Add missing servicenow plugin checks

EdwardDowling committed:
Make omit empty fields in servicenow incident type

EdwardDowling committed:
Enforce https for servicenow requests

EdwardDowling committed:
Make servicenow plugin always pass clustername

EdwardDowling committed:
Add in missing returned errors in servicenow plugin

EdwardDowling committed:
Remove unused code form servicenow plugin

EdwardDowling committed:
Add missing godocs to servicenow types

EdwardDowling committed:
Use parseint over Sscanf in decodeplugindata

EdwardDowling committed:
Fix minor formating issues in servicenow bot

EdwardDowling committed:
Fix min version for servicenow plugin

EdwardDowling committed:
Rename servicenow to serviceNow

EdwardDowling committed:
Add more context to servicenow logs and send time as utc

EdwardDowling committed:
Fix misleading comment in servicenow plugin

EdwardDowling committed:
Add more context to servicenow plugin logging

EdwardDowling commented on integrations/access/servicenow/client.go:
```diff
@@ -223,6 +223,24 @@ func (snc *Client) GetOnCall(ctx context.Context, rotaID string) ([]string, erro
+		SetQueryParams(map[string]string{
+			"sysparm_limit": "1",
+		}).
+		Get("/api/now/table/incident")
```
The error when that table isn't present would be included in the errwrapper section. I could include the requirement for the table in the docs?

marcoandredinis replied:
Do you know what the error message would look like?
If it's clear enough, than it's ok to leave as is
For the docs, I think it would be worth it to add a note (I guess that's another PR)

Also, I recall we have a plugin status field
Should this kind of error trigger a change in the status field to indicate to the user that this is not working (and why/how to fix it)
No need to do this now, just something that I think we should consider

EdwardDowling replied:
So do you mean using the plugin status field for errors like this where they can't be retried? Would that be  clearer than just logging those cases?

marcoandredinis replied:
Yeah, I think so
Assuming this is going to be run first in cloud hosted envs, customers won't have access to logs

EdwardDowling replied:
That sounds good to me then. Could throw that in with the PR that will be adding the auto approval flow?

marcoandredinis replied:
Sure :+1: 
Or a PR just to handle that scenario

EdwardDowling committed:
Refactor servicenow plugin to reduce indentation

marcoandredinis commented on integrations/access/servicenow/plugindata.go:
```diff
@@ -0,0 +1,77 @@
+	}
+	data.RequestReason = dataMap["request_reason"]
+	if str := dataMap["reviews_count"]; str != "" {
+		fmt.Sscanf(str, "%d", &data.ReviewsCount)
```
Should we also use `strconv.ParseInt` here?

marcoandredinis commented on integrations/access/servicenow/plugindata.go:
```diff
@@ -0,0 +1,77 @@
+
+	var createdStr string
+	if !data.Created.IsZero() {
+		createdStr = fmt.Sprintf("%d", data.Created.Unix())
```
:nail_care: 
Should we use `strconv.Itoa` ? (the same applies for the ReviewsCount)

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,607 @@
+		resolution.State = ResolutionStateResolved
+	}
+
+	// TODO: Add in close codes
```
Is this something we will do in a future PR?

marcoandredinis reviewed (commented):
I saw your loom recording for this new feature :clap: 

Would you mind adding some tests

EdwardDowling committed:
Refactor resolution tag check to use a switch case

EdwardDowling commented on integrations/access/servicenow/bot.go:
```diff
@@ -0,0 +1,121 @@
+	var closeCode string
+	var state string
+	if data.ResolutionTag == pd.ResolvedApproved {
+		values, ok := data.SystemAnnotations[types.TeleportNamespace+types.ReqAnnotationApprovedCloseCode]
```
This is added to the role that is making the access requests.

EdwardDowling committed:
Change from sscanf anf sprintf for int formating and parsing

EdwardDowling commented on integrations/access/servicenow/client.go:
```diff
@@ -85,6 +85,11 @@ func NewClient(conf ClientConfig) (*Client, error) {
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	apiURL.Scheme = "https"
```
Would making it so the config has a 'scheme' field that defaults to 'https' and is only set in the tests to avoid this be ok?

marcoandredinis replied:
Sure, either that (+guarded by only allowing http on localhost) or passing in the CA from the test server

If going with the latter, I did something similar here
https://github.com/gravitational/teleport/pull/30632/commits/8e4191abf52fff78993a393d0701593126a0e7f2

EdwardDowling replied:
How about this instead? No need for config fields that would only be used in testing this way.

EdwardDowling committed:
Add missing close code logic to servicenow bot

EdwardDowling commented on integrations/access/servicenow/bot.go:
```diff
@@ -0,0 +1,119 @@
+	if data.ResolutionTag == pd.ResolvedApproved {
+		values, ok := data.SystemAnnotations[types.TeleportNamespace+types.ReqAnnotationApprovedCloseCode]
+		if !ok || len(values) < 1 {
+			return trace.BadParameter("close code annotation missing form serviceNow configuration")
```
It could also be the case they simply didnt add the close code to that requesting role.

tigrato commented on api/types/constants.go:
```diff
@@ -662,6 +662,11 @@ const (
 	ReqAnnotationNotifyServicesLabel = "/notify-services"
 
+	// ReqAnnotationApprovedCloseCode is the request annotation key at which the servicenow close code for approved access requests is stored, this label is added to roles to determine what close code to use when closing the ServiceNow incidents created for access requests.
+	ReqAnnotationApprovedCloseCode = "/approved-close-code"
```
I still see that you check `req.GetSystemAnnotations()[types.TeleportNamespace+types.ReqAnnotationDeniedCloseCode]` but I don't see where you set the system annotations at all.
So, I guess that something is still missing somewhere to set this value

EdwardDowling replied:
The current idea for this is that these are set in the requesting role

tigrato commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,625 @@
+		return nil
+	}
+	// Don't show the error if the annotation is just missing.
+	if trace.Unwrap(notifyErr) == errMissingAnnotation {
```
```suggestion
	if errors.Is(notifyErr, errMissingAnnotation) {
```

tigrato commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,625 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	userIsOnCall := false
+	for _, user := range onCallUsers {
+		if req.GetUser() == user {
+			userIsOnCall = true
+		}
+	}
```
```suggestion
	userIsOnCall := slices.Contains(onCallUsers, req.GetUser())
```

EdwardDowling committed:
Add tests for servicenow plugin and mocked servicenow

EdwardDowling committed:
Refactor servicenow plugin to use slice.Contains

EdwardDowling committed:
Enforce https only for non localhost in servicenow plugin

EdwardDowling committed:
Add more context to servicenow close code comments

tigrato reviewed (approved):
Please document where the labels are coming from, otherwise LGTM

marcoandredinis commented on api/types/plugin.go:
```diff
@@ -506,6 +522,14 @@ func (c *PluginDiscordSettings) CheckAndSetDefaults() error {
 	return nil
 }
 
+func (c *PluginServiceNowSettings) CheckAndSetDefaults() error {
```
Missing go docs

marcoandredinis commented on integrations/access/servicenow/bot.go:
```diff
@@ -0,0 +1,120 @@
+	case pd.ResolvedApproved:
+		values, ok := data.SystemAnnotations[types.TeleportNamespace+types.ReqAnnotationApprovedCloseCode]
+		if !ok || len(values) < 1 {
+			return trace.BadParameter("close code annotation missing form serviceNow configuration")
```
```suggestion
			return trace.BadParameter("close code annotation missing form ServiceNow configuration")
```

marcoandredinis commented on integrations/access/servicenow/bot.go:
```diff
@@ -0,0 +1,120 @@
+	case pd.ResolvedDenied:
+		values, ok := data.SystemAnnotations[types.TeleportNamespace+types.ReqAnnotationDeniedCloseCode]
+		if !ok || len(values) < 1 {
+			return trace.BadParameter("close code annotation missing form serviceNow configuration")
```
```suggestion
			return trace.BadParameter("close code annotation missing form ServiceNow configuration")
```

EdwardDowling committed:
Fix typo in error message and encode systemAnnotations in plugindata

marcoandredinis commented on integrations/access/servicenow/client.go:
```diff
@@ -85,6 +85,18 @@ func NewClient(conf ClientConfig) (*Client, error) {
+		return nil, trace.Wrap(err)
+	}
+
+	if apiURL.Scheme == "http" && !strings.HasPrefix(apiURL.Host, "127.0.0.1") {
```
Can you do an exact match against `apiURL.Hostname()` instead? And compare against `127.0.0.1` or `localhost`.

This currently returns true for hosts like `127.0.0.1.evil.com`

EdwardDowling replied:
Something like this?

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,621 @@
+
+		var err error
+		switch {
+		case req.GetState().IsPending():
+			err = a.onPendingRequest(ctx, req)
+		case req.GetState().IsApproved():
+			err = a.onResolvedRequest(ctx, req)
+		case req.GetState().IsDenied():
+			err = a.onResolvedRequest(ctx, req)
```
Can we replace the `IsApproved` and `IsDenied` (and merge the case branch) with `IsResolved`. Should be the same thing for now.

There's an ongoing PR that adds a new state.
https://github.com/gravitational/teleport/pull/31346/files
This way seems safer: less probability of someone forgetting to consider the new state

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,621 @@
+	case types.RequestState_APPROVED:
+		values, ok := req.GetSystemAnnotations()[types.TeleportNamespace+types.ReqAnnotationApprovedCloseCode]
+		if !ok || len(values) < 1 {
+			return trace.BadParameter("close code annotation missing form serviceNow configuration")
```
```suggestion
			return trace.BadParameter("close code annotation missing form ServiceNow configuration")
```

marcoandredinis commented on integrations/access/servicenow/app.go:
```diff
@@ -0,0 +1,621 @@
+	case types.RequestState_DENIED:
+		values, ok := req.GetSystemAnnotations()[types.TeleportNamespace+types.ReqAnnotationDeniedCloseCode]
+		if !ok || len(values) < 1 {
+			return trace.BadParameter("close code annotation missing form serviceNow configuration")
```
```suggestion
			return trace.BadParameter("close code annotation missing form ServiceNow configuration")
```

marcoandredinis commented on integrations/access/servicenow/servicenow_test.go:
```diff
@@ -0,0 +1,556 @@
+
+	// We set such a big timeout because integration.NewFromEnv could start
+	// downloading a Teleport *-bin.tar.gz file which can take a long time.
+	ctx := s.SetContextTimeout(1 * time.Minute)
```
Is this still required?

EdwardDowling replied:
Ah, will remove it.

EdwardDowling committed:
Match host for servicenow client as whole instead of via prefix

And fix capitalisation in some error messages.

EdwardDowling committed:
Merge denied and approved cases in servicenow on watcher event

EdwardDowling committed:
Lower excessive timeout in servicenow tests

marcoandredinis commented on integrations/access/servicenow/client.go:
```diff
@@ -90,10 +91,11 @@ func NewClient(conf ClientConfig) (*Client, error) {
 	}
 
-	if apiURL.Scheme == "http" && !strings.HasPrefix(apiURL.Host, "127.0.0.1") {
+	hostNoPort, _, err := net.SplitHostPort(apiURL.Host)
```
You are not checking the `err` value.
But you can use `apiURL.Hostname()` to get the same result

EdwardDowling replied:
Oh nice. Didn't know that existed.

EdwardDowling committed:
Swap to using apiurl.hostname of net.splithostport

r0mant commented on api/types/constants.go:
```diff
@@ -662,6 +662,11 @@ const (
 	ReqAnnotationNotifyServicesLabel = "/notify-services"
 
+	// ReqAnnotationApprovedCloseCode is the request annotation key at which the servicenow close code for approved access requests is stored, this label is added to roles to determine what close code to use when closing the ServiceNow incidents created for access requests. ServiceNow close codes are set on the requesting role.
+	ReqAnnotationApprovedCloseCode = "/approved-close-code"
+	// ReqAnnotationDeniedCloseCode is the request annotation key at which the servicenow close code for denied access requests is stored, this label is added to roles to determine what close code to use when closing the ServiceNow incidents created for access requests. ServiceNow close codes are set on the requesting role.
+	ReqAnnotationDeniedCloseCode = "/denied-close-code"
```
@EdwardDowling Can you clarify what these are for? Why would a user need to set these on the roles? Seems confusing.

EdwardDowling replied:
Changing this so close code will be configured at the plugin level instead of per role

r0mant commented on integrations/access/common/recipient.go:
```diff
@@ -26,6 +26,8 @@ import (
 const (
 	// RecipientKindSchedule shows a recipient is a schedule.
 	RecipientKindSchedule = "schedule"
+	// RecipientKindRota shows a recipient is a rota.
```
What's "rota"? Rotation? Let's just say so.

r0mant reviewed (changes requested):
Requesting changes for now until we clarify what these "close code" annotations are for.

EdwardDowling commented on integrations/access/servicenow/client.go:
```diff
@@ -85,6 +85,18 @@ func NewClient(conf ClientConfig) (*Client, error) {
+		return nil, trace.Wrap(err)
+	}
+
+	if apiURL.Scheme == "http" && apiURL.Hostname() != "127.0.0.1" {
+		return nil, trace.BadParameter("http scheme is only permitted for localhost: %v", apiURL.Host)
+	}
```
This is just for the tests.

EdwardDowling committed:
Fix access_request plugindata test, rename servicenow rota

EdwardDowling committed:
Move close code configuration to plugin instead of role

EdwardDowling committed:
Remove unused fields and fix minor linting issues

EdwardDowling commented:
> Requesting changes for now until we clarify what these "close code" annotations are for.

Added in those changes we discussed on slack to this and the /e PR.
Moved the close code configuration to be at the plugin level instead of per role.

EdwardDowling committed:
Remove unused code to fix linting

EdwardDowling committed:
Remove unused assignment to fix linting

EdwardDowling committed:
Merge branch 'master' into edwarddowling/service-now-plugin

EdwardDowling committed:
Fix assignment error in servicenow tests
